crates/wdl-engine/src/cache/hash.rs

```diff
@@ -0,0 +1,320 @@
+//! Implementation of hashing values for the call cache.
+
+use std::path::Path;
+use std::path::PathBuf;
+
+use blake3::Hasher;
+use url::Url;
+
+use crate::Array;
+use crate::CompoundValue;
+use crate::HiddenValue;
+use crate::HintsValue;
+use crate::InputValue;
+use crate::Map;
+use crate::Object;
+use crate::OutputValue;
+use crate::Pair;
+use crate::PrimitiveValue;
+use crate::Struct;
+use crate::Value;
+use crate::digest::Digest;
+use crate::path::EvaluationPath;
+
+/// The variant tag for `None` values.
+const NONE_VARIANT_TAG: u8 = 0;
+/// The variant tag for `Boolean` values.
+const BOOLEAN_VARIANT_TAG: u8 = 1;
+/// The variant tag for `Int` values.
+const INTEGER_VARIANT_TAG: u8 = 2;
+/// The variant tag for `Float` values.
+const FLOAT_VARIANT_TAG: u8 = 3;
+/// The variant tag for `String` values.
+const STRING_VARIANT_TAG: u8 = 4;
+/// The variant tag for `File` values.
+const FILE_VARIANT_TAG: u8 = 5;
+/// The variant tag for `Directory` values.
+const DIRECTORY_VARIANT_TAG: u8 = 6;
+/// The variant tag for `Pair` values.
+const PAIR_VARIANT_TAG: u8 = 7;
+/// The variant tag for `Array` values.
+const ARRAY_VARIANT_TAG: u8 = 8;
+/// The variant tag for `Map` values.
+const MAP_VARIANT_TAG: u8 = 9;
+/// The variant tag for `Object` values.
+const OBJECT_VARIANT_TAG: u8 = 10;
+/// The variant tag for `Struct` values.
+const STRUCT_VARIANT_TAG: u8 = 11;
+/// The variant tag for `hints` values.
+const HINTS_VARIANT_TAG: u8 = 12;
+/// The variant tag for `input` values.
+const INPUT_VARIANT_TAG: u8 = 13;
+/// The variant tag for `output` values.
+const OUTPUT_VARIANT_TAG: u8 = 14;
+
+/// The variant tag for local evaluation paths.
+const LOCAL_PATH_VARIANT_TAG: u8 = 0;
+/// The variant tag for remote evaluation paths.
+const REMOTE_PATH_VARIANT_TAG: u8 = 1;
+
+/// The variant tag for hash-based remote content digests.
+const CONTENT_HASH_VARIANT_TAG: u8 = 0;
+/// The variant tag for ETag based remote content digests.
+const CONTENT_ETAG_VARIANT_TAG: u8 = 1;
+
+/// The variant tag for file digests.
+const DIGEST_FILE_VARIANT_TAG: u8 = 0;
+/// The variant tag for directory digests.
+const DIGEST_DIRECTORY_VARIANT_TAG: u8 = 1;
```

Comment:
The [`num_enum` crate](https://docs.rs/num_enum/0.7.5/num_enum/) might be nice to use here.